Experiments: sharing private APIs with lock() and unlock()

adamziel opened the pull request:
## Description

This commit introduces a more convenient API for managing the private experiments as the former `register`-based API was quite cumbersome to use.

The idea is to "lock" private data inside public objects:

```js
const { lock, unlock } = __dangerousOptInToUnstableAPIsOnlyForCoreModules(
	'<CONSENT STRING>',
	'@wordpress/blocks'
);

export const publicObject = {};
lock( __experiments, "Shh, private data!" );

publicObject
// {}

unlock( publicObject )
// "Shh, private data!"
```

This new `lock()`/`unlock()` API enables private functions, classes, components, selectors, actions, arguments, and properties. Any package that opted-in to private APIs can call `unlock()` on  publicly available artifacts to retrieve the related private API.

Kudos to @jsnajdr for [identifying an opportunity to simplify the API](https://github.com/WordPress/gutenberg/pull/44521#discussion_r983306757)!

## Examples

### Private selectors:

```js
// In wordpress/block-data:
import { store as blockEditorStore } from './store';
import { unlock } from '../experiments';
import { __unstableSelectionHasUnmergeableBlock } from './store/selectors';

unlock( store ).registerPrivateSelectors( {
  __unstableSelectionHasUnmergeableBlock
} );

// In a React component:
function MyComponent() {
    const hasRole = useSelect( ( select ) => (
        unlock( select( blockEditorStore ) ).__unstableSelectionHasUnmergeableBlock()
    ) );

    // ...
}
```

### Private functions, classes, and variables

```js
// In packages/package1/index.js:
import { lock } from './experiments';

export const experiments = {};
/* Attach private data to the exported object */
lock(experiments, {
	__experimentalCallback: function() {},
	__experimentalReactComponent: function ExperimentalComponent() { return <div/>; },
	__experimentalClass: class Experiment{},
	__experimentalVariable: 5,
});


// In packages/package2/index.js:
import { experiments } from '@wordpress/package1';
import { unlock } from './experiments';

const {
	__experimentalCallback,
	__experimentalReactComponent,
	__experimentalClass,
	__experimentalVariable
} = unlock( experiments );
```

### Private function arguments

To add an experimental argument to a stable function you'll need
to prepare a stable and an experimental version of that function.
Then, export the stable function and `lock()` the unstable function
inside it:

```js
// In @wordpress/package1/index.js:
import { lock } from './experiments';

// The experimental function contains all the logic
function __experimentalValidateBlocks(formula, __experimentalIsStrict) {
	let isValid = false;
	// ...complex logic we don't want to duplicate...
	if ( __experimentalIsStrict ) {
		// ...
	}
	// ...complex logic we don't want to duplicate...

	return isValid;
}

// The stable public function is a thin wrapper that calls the
// experimental function with the experimental features disabled
export function validateBlocks(blocks) {
	__experimentalValidateBlocks(blocks, false);
}
lock( validateBlocks, __experimentalValidateBlocks );

// In @wordpress/package2/index.js:
import { validateBlocks } from '@wordpress/package1';
import { unlock } from './experiments';

// The experimental function may be "unlocked" given the stable function:
const __experimentalValidateBlocks = unlock(validateBlocks);
__experimentalValidateBlocks(blocks, true);
```

### Private React Component properties

To add an experimental argument to a stable component you'll need
to prepare a stable and an experimental version of that component.
Then, export the stable function and `lock()` the unstable function
inside it:

```js
// In @wordpress/package1/index.js:
import { lock } from './experiments';

// The experimental component contains all the logic
const ExperimentalMyButton = ( { title, __experimentalShowIcon = true } ) => {
	// ...complex logic we don't want to duplicate...
  
	return (
		<button>
			{ __experimentalShowIcon  && <Icon src={some icon} /> } { title }  
		</button>
	);
}

// The stable public component is a thin wrapper that calls the
// experimental component with the experimental features disabled
export const MyButton = ( { title } ) => 
    <ExperimentalMyExistingButton title={ title } __experimentalShowIcon={ false } />

lock(MyButton, ExperimentalMyButton);


// In @wordpress/package2/index.js:
import { MyButton } from '@wordpress/package1';
import { unlock } from './experiments';

// The experimental component may be "unlocked" given the stable component:
const ExperimentalMyButton = unlock(MyButton);
export function MyComponent() {
	return (
		<ExperimentalMyButton data={data} __experimentalShowIcon={ true } />
	)
}
```

## Testing Instructions

Confirm the tests pass. Comment with your opinion on this API.

cc @dmsnell @gziolo @jsnajdr @jorgefilipecosta @priethor @georgeh @artemiomorales @tellthemachines @draganescu @peterwilsoncc

adamziel committed:
Experiments: pivot towards lock()/unlock() based API

This commit introduces a more convenient API for managing the
private experiments.

The idea is to "lock" private data inside public objects:

```js
const { lock, unlock } = __dangerousOptInToUnstableAPIsOnlyForCoreModules(
	'<CONSENT STRING>',
	'@wordpress/blocks'
);

export const publicObject = {};
lock( __experiments, "Shh, private data!" );

publicObject
// {}

unlock( publicObject )
// "Shh, private data!"
```

The lock()/unlock() API makes it easier to implement private functions, components,
selectors, actions etc. Instead of using a separate import system, it's enough to
opt-in to the experiments and then `unlock()` the publicly available
artifact. For example:

```js
import { store } from '@wordpress/blocks';
import { useSelect } from '@wordpress/data';
import { unlock } from './experiments';

function MyComponent() {
    const hasRole = useSelect( ( select ) => (
        unlock( select( store ) ).__experimentalHasContentRoleAttribute()
    ) );

    // ...
}
```

This makes `unlock()` a drop-in replacement that does not require
reworking the current code structure around the experimental APIs.

What's also nice is that the above implementation of private selectors
(coming in a follow-up PR) is short, sweet, and fairly easy to grasp.
For comparison, an implementation based on the former register/unlock
API required quite a bit of mental gymnastics:

```js
// Package wordpress/block-editor:
export const { register, unlock } =
	__dangerousOptInToUnstableAPIsOnlyForCoreModules(
		'I know using unstable features means my plugin or theme will inevitably break on the next WordPress release.',
		'@wordpress/block-editor'
	);

const { __experimentalPrivateSelector } = unlock( dataExperiments );

import {
	__unstableSelectionHasUnmergeableBlock,
} from './store/selectors';

export const __experimentalAccessKey = register( {
	__unstableSelectionHasUnmergeableBlock: __experimentalPrivateSelector(
		store,
		__unstableSelectionHasUnmergeableBlock
	)
} );

// Package wordpress/edit-page
export const { register, unlock } =
         __dangerousOptInToUnstableAPIsOnlyForCoreModules(
                 'I know using unstable features means my plugin or theme will inevitably break on the next WordPress release.',
                 '@wordpress/block-editor'
         );

import { __experimentalAccessKey as blockEditorExperiments } from '@wordpress/block-editor';
const { __unstableSelectionHasUnmergeableBlock } = unlock(__experimentalAccessKey);
```

It also had an undesirable side-effect of requiring a separate import
and a different treatment of the `__unstableSelectionHasUnmergeableBlock()`
selector – which is to say a migration would be rather difficult.

adamziel committed:
Rename lazyDecorator to onFirstUnlock and use it to process even the null values

adamziel committed:
Document the new API

youknowriad commented on packages/experiments/README.md:
```diff
@@ -7,33 +7,42 @@ This package acts as a "dealer" that only allows WordPress packages to use the e
+// In the package consuming the experimental APIs:
+const { unlock } = __dangerousOptInToUnstableAPIsOnlyForCoreModules(
+	'<CONSENT STRING>', // See index.js, this may change without notice.
+	'@wordpress/block-editor'
```
I'm probably not understanding how this work but I thought that to "unlock" '@wordpress/blocks' APIs, I need to provide the same package name here?

adamziel replied:
@youknowriad The code you highlighted is a deterrent to only deal the `lock()` and `unlock()` functions to core WordPress packages. By passing the `@wordpress/block-editor` package name, I'm saying "This is `@wordpress/block-editor`" opting in, do not accept this name again". In other words, I couldn't do this:

```js
// Only the first opt-in for a given package name works:
_dangerousOptInToUnstableAPIsOnlyForCoreModules(
	'<CONSENT STRING>',
	'@wordpress/block-editor'
);

// Any subsequent attempt throws an exception:
_dangerousOptInToUnstableAPIsOnlyForCoreModules(
	'<CONSENT STRING>',
	'@wordpress/block-editor'
);
```

adamziel replied:
The idea is to hide private data in public objects via `lock()`, and retrieve it via `unlock()` as follows:

```js
// In @wordpress/package1:
export const publicObject = {}; 
lock(publicObject, 'Private string');

// In @wordpress/package2:
import { publicObject } from '@wordpress/package1';

console.log(unlock(publicObject));
// "Private string"

console.log(publicObject);
// Just "{}", you can't directly retrieve the private string from that object
```

So you don't have to provide the source package name when unlocking. Instead, you import an object to unlock from that package – this makes the cross-package dependencies clear and prevents you from quietly using `@wordpress/package1` export without listing it in package.json.

youknowriad commented on packages/experiments/src/implementation.js:
```diff
@@ -0,0 +1,279 @@
+	'@wordpress/data',
+	'@wordpress/blocks',
+	'@wordpress/block-editor',
+];
```
Why is there a fixed list of packages here?

adamziel replied:
@youknowriad see my [other comment](https://github.com/WordPress/gutenberg/pull/46131#discussion_r1052298567) – this is an allowlist of which packages are allowed to access to `lock()` and `unlock()` functions.

youknowriad replied:
A bit weird though, if one of these package is not loaded on a given page, a third-party tool can still use the said API.

adamziel replied:
There is no watertight solution here, unfortunately. Deterrents are as close as it gets: the existence of the `unlock()` API, the consent string, and the packages allowlist are designed to 1) make it impossible to use private APIs accidentally 2) communicate the consequences and 3) make your life harder if you still try to go for it.

Importantly, there's buy-in that this method should enable Gutenberg to smoothly ship and remove private cross-package experiments even after a major WordPress core release. It's hard to claim your project is broken because WordPress failed to keep the BC promise when your code literally says you knew that would happen.

youknowriad replied:
Let's keep the discussion on this particular point please. (For the buy in, I'd say it's not complete to say the least)

adamziel replied:
On that particular point it's as I said – yes, you can work around it with enough effort. That effort, plus making your code even more likely to break, is the point.

youknowriad commented on packages/experiments/README.md:
```diff
@@ -7,33 +7,42 @@ This package acts as a "dealer" that only allows WordPress packages to use the e
-
-const { __unstableGetInnerBlocksProps } = unlock( blocksExperiments );
+// In the package exposing the experimental APIs:
+const { lock } = __dangerousOptInToUnstableAPIsOnlyForCoreModules(
```
Can you share an example with a package exporting an experimental component?

adamziel replied:
Sure!

```js
// In @wordpress/package1/index.js:
import { lock } from './experiments';

export function DataListing({data}) {
	return <ExperimentalDataListing data={data} format='list' />;
}
function ExperimentalDataListing({data, format}) {
	if(format === 'list') {
		return (<ul> {/* ... */} </ul>)
	} else if(format === 'table') {
		return (<table> {/* ... */} </table>)
	}
}

// In @wordpress/package2/index.js:
import { DataListing } from '@wordpress/package1';
import { unlock } from './experiments';

const ExperimentalDataListing = unlock(DataListing);
export function MyComponent() {
	return (
		<ExperimentalDataListing data={data} format="table" />
	)
}
```

Without the `unlock()` function you can only use the public `DataListing` component. With the `unlock()` access you can also access the private `ExperimentalDataListing` component.

mcsf replied:
If I am understanding the proposal correctly, there is a missing `lock( DataListing, ExperimentalDataListing )` call in your example, right?

youknowriad reviewed (commented):
Just noting that unless I'm missing something, this approach won't work for props, function arguments, keys in block.json, theme.json.

adamziel committed:
Add unit tests to demonstrate how to export private functions, function arguments, and react components

adamziel committed:
Explain how to avoid new experimental APIs in coding-guidelines.md

adamziel commented:
> Just noting that unless I'm missing something, this approach won't work for props, function arguments, keys in block.json, theme.json.

@youknowriad it will work for anything you can export – see my comments. I also added some examples to the contributing docs.

As for the `block.json` and `theme.json` keys – you're right. We'd need to preprocess the `block.json` and `theme.json` to remove the `__experimental` keys for third party blocks and themes. This preprocessor would only remove the experiments merged **later than the preprocessor**. All the existing public `__experimental` keys would continue to work for the existing plugins and themes, but no new `__experimental` key would. I'm playing with that on a local branch. What do you think?

adamziel commented:
Also CCing @jsnajdr for comments

youknowriad commented on docs/contributors/code/coding-guidelines.md:
```diff
@@ -136,10 +136,278 @@ export { __unstableDoTerribleAwfulAction } from './api';
+	// MyComponent.js:
+	function MyComponent() {
+		const registry = useRegistry();
+		const onClick = () => registry.stores['my-store'].store.dispatch({ type: 'BEFORE_SAVE' });
```
Actually, this is a bad approach, we should always be using useDispatch to access actions. This stores key has always been meant as an internal variable. I'd remove this example from here.

youknowriad replied:
The only reason this "stores" key exist is to make other public APIs work like "use" (which I believe we should deprecate)

adamziel replied:
Ah thank you, let's remove it indeed then

adamziel replied:
> I don't think it solves the problem for selectors, actions (...)

@youknowriad note this doc fragment discusses _replacing experimental actions with hooks and thunks_. I'll remove it, but it does not affect the section about `lock()` and `unlock()` below:

```
unlock( select( store ) ).__experimentalHasContentRoleAttribute()
```

youknowriad commented on docs/contributors/code/coding-guidelines.md:
````diff
@@ -136,10 +136,278 @@ export { __unstableDoTerribleAwfulAction } from './api';
+unlock(log)('table');
+```
+
+#### Experimental React Component properties
````
I also think this is misleading, as adding properties often (if not always) need to impact the internals of components, so you do a condition to render two separate trees unless you duplicate a lot of code.  Same for the functions.

adamziel replied:
@youknowriad Whenever you have an experimental property/argument, you also need a well-defined stable behavior. The pattern of "a full-featured private component" and "limited public facade enforcing the stable behavior" captures that dynamics. Do you have any counter-examples where that wouldn't work?

youknowriad replied:
I'd say most cases where we do experimental props don't work here (unless I'm missing something). You just add a prop and you want to change something in the rendered elements, example:

```js
const MyExistingButton = ( { title } ) => {
  // Do some complex stuff that I don't want to duplicate.
  
  return (
       <button>
           <Icon src={some icon} /> { title }  
      </button>
    );
}

```

now introduce the experimental prop

```
const MyExistingButton = ( { title, __experimentalShowIcon = true } ) => {
  // Do some complex stuff that I don't want to duplicate.
  
  return (
       <button>
           { __experimentalShowIcon  && <Icon src={some icon} /> } { title }  
      </button>
    );
}
```

How do I introduce "__experimentalShowIcon" without duplicating the whole component?

adamziel replied:
@youknowriad Here's what I'd do:

```js
export const MyExistingButton = ( { title } ) => 
    <ExperimentalMyExistingButton title={ title } __experimentalShowIcon={ false } />

const ExperimentalMyExistingButton = ( { title, __experimentalShowIcon = true } ) => {
  // Do some complex stuff that I don't want to duplicate.
  
  return (
       <button>
           { __experimentalShowIcon  && <Icon src={some icon} /> } { title }  
      </button>
    );
}

lock(MyExistingButton, ExperimentalMyExistingButton);
```

Actually, if `__experimentalShowIcon` defaulted to false, we could do just:

```js
export const MyExistingButton = ( { title } ) => 
    <ExperimentalMyExistingButton title={ title } />
```

The public component does not accept any experimental props. The private component does accepts experimental props. The public component calls the private one such that the experimental features are disabled.

youknowriad replied:
Makes sense, I didn't think about inverting the responsibilities.

youknowriad replied:
It still feels weird though to have to define a second component to be able to add an argument to an existing one.

youknowriad commented on docs/contributors/code/coding-guidelines.md:
```diff
@@ -136,10 +136,278 @@ export { __unstableDoTerribleAwfulAction } from './api';
+function MyComponent() {
+    const hasRole = useSelect( ( select ) => (
+		// Use the private selector:
+        unlock( select( store ) ).__experimentalHasContentRoleAttribute()
```
My understanding is that `unlock` can only be called once for a given package, this is going to be called several times, is it going to work still? If yes, I don't see how this stops any third party code from accessing these selectors/actions.

adamziel replied:
You can call `unlock()` as many times as you want. The catch is the only way to access the `unlock` function is by calling the `__dangerousOptInToUnstableAPIsOnlyForCoreModules` function:

```js
// packages/block-editor/experiments.js
import { __dangerousOptInToUnstableAPIsOnlyForCoreModules } from '@wordpress/experiments';
const { lock, unlock } = __dangerousOptInToUnstableAPIsOnlyForCoreModules(
	'<CONSENT STRING>',
	'@wordpress/blocks'
);

// packages/block-editor/my-component.js
import { unlock } from './experiments';

unlock( select( store ) ). __experimentalHasContentRoleAttribute()
unlock( select( store ) ). __experimentalHasContentRoleAttribute()
```

I'll describe it more explicitly

youknowriad replied:
Ok makes sense, I guess that means "selectors and actions" are considered solved by this API 👍 (unlike props, arguments, theme.json, block.json)

youknowriad reviewed (commented):
I like how this works for exported functions, components but I don't think it solves the problem for selectors, actions, props, function arguments, theme.json, block.json yet. (see my comments)

In other words, for now at least, I'm ok with this being the approach to take for exported functions and components but until we figure out the rest, we should be able to continue with our current approach. (would love if the documentation is updated as such accordingly)

adamziel commented:
@youknowriad I'll update the documentation and I agree about `block.json` and `theme.json`. I have some notes about the selectors, actions, props, and function arguments, though – see the inline comments.

adamziel committed:
Editorial updates to the documentation

youknowriad commented on docs/contributors/code/coding-guidelines.md:
````diff
@@ -176,9 +176,11 @@ if ( IS_GUTENBERG_PLUGIN ) {
 }
 ```
 
+The experimental APIs that already are publicly available via `window.wp` should remain accessible as removing them would break the existing plugins and themes.
````
I'm not sure I agree with this personally. I think for the time being, we should continue with our current policies that we've been using for some time now. I think we can deprecate these APIs and remove them once we know they don't have any impact:

 - Deprecating
 - Giving time,
 - On timeout, checking usage and remove

adamziel replied:
Hm yes – good note. I updated the document.

youknowriad reviewed (changes requested):
I think this looking good, only thing for me is that I don't think we should be changing the policy for existing APIs and should remove this line. Thanks https://github.com/WordPress/gutenberg/pull/46131/files/c541a2c807a3a7e8af2f718e0ae99dabdb411829..05aac1d97a5ccce1cd5dac449db7593a68eee2d4#r1054431055

mcsf commented on packages/experiments/src/implementation.js:
```diff
@@ -0,0 +1,279 @@
+ * @param {Function} [config.onFirstUnlock] A function to run on the unlocked data when `unlock()`
+ *                                          is called for the first time.
+ */
+export function configureExperiment( object, config ) {
```
I can't remember if I've already asked this (maybe even in person?) — where do we intend to use this?

mcsf replied:
Same question for the configuration abstraction (`getLockingConfig`, `isExperimentsConfig`, etc.).

adamziel replied:
@mcsf Your comment tells me the documentation strings aren't very clear – I'll update them.

There's another PR against this branch where these utilities are used to enable registering private selectors and actions: https://github.com/WordPress/gutenberg/pull/44521

In short, `lock( store, { selectors: {} } )` affects the store but `useSelect( store )` does not return the store object so these utilities make `unlock( useSelect( store ) )` return the private selectors tied to the store.

adamziel replied:
I'd love to merge https://github.com/WordPress/gutenberg/pull/44521 into this one before merging this into trunk.

adamziel replied:
@mcsf I merged https://github.com/WordPress/gutenberg/pull/44521 into this one to make the big picture clear and also simplified the exported API. I hope it's less confusing now.

adamziel committed:
Editorial updates to coding guidelines

youknowriad commented on docs/contributors/code/coding-guidelines.md:
```diff
@@ -136,10 +136,306 @@ export { __unstableDoTerribleAwfulAction } from './api';
+features that can't be accessed in any other way. Naturally, these APIs can't be removed without a warning anymore.
+They are a part of the WordPress public API and fall under the 
+[WordPress Backwards Compatibility policy](https://developer.wordpress.org/block-editor/contributors/code/backward-compatibility/). 
+Removing them involves a deprecation process that spans multiple WordPress releases. 
```
I think the length really depends on the API, some are not useful or not used at all by plugins and should be easier to remove, others are less so.

adamziel replied:
Good note, I rewrote this to note both extremes (fairly straightforward vs hard and long)

youknowriad reviewed (commented):
No blockers from me any more. Thanks

adamziel committed:
Do not export isExperimentsConfig from the experiments package

adamziel committed:
Clarify the deprecation process does not always span multiple WP
releases

youknowriad reviewed (approved):
I'd love more opinions and thoughts about this PR as it's a departure from how we do things. So @WordPress/gutenberg-core should be aware of this.

Thanks for pushing through this.

adamziel committed:
Private experimental cross-module selectors and actions (#44521)

Introduces a private selectors APIs in `@wordpress/data` via [the new `@wordpress/experimental`](https://github.com/WordPress/gutenberg/pull/43386#issuecomment-1260379417) package:

```js
// Package wordpress/block-data:
import { unlock } from '../experiments';
import { experiments as dataExperiments } from '@wordpress/data';
const { registerPrivateActionsAndSelectors } = unlock( dataExperiments );

import { store as blockEditorStore } from './store';
import { __unstableSelectionHasUnmergeableBlock } from './store/selectors';
registerPrivateActionsAndSelectors( store, {}, {
     __experimentalHasContentRoleAttribute
} );

// plain usage
unlock( registry.select( blockEditorStore ) ).getContentLockingParent();

// usage in React
useSelect( ( select ) => ( {
  parent: privateOf( select( blockEditorStore ) ).__unstableSelectionHasUnmergeableBlock();
} ) );
```

adamziel committed:
Roll back the demo changes to blocks/ and block-editor/ packages

adamziel committed:
Update the coding guidelines to reflect registerPrivateActionsAndSelectors()

adamziel committed:
Lint

adamziel commented:
Also cc @micbosi @jsnajdr @dmsnell @peterwilsoncc @gziolo @aaronrobertshaw @andrewserong

adamziel committed:
Use consistent method names in docstring examples

mcsf commented:
I'm going to comment at once on both this PR and its tributary https://github.com/WordPress/gutenberg/pull/44521.

First of all, I think this is the most solid _code-based_ solution to the API problem so far, so nice work there. It's relatively comprehensive in the kind of APIs it could absorb, and flexible in how to manage locked objects. That said, I agree with Riad that it is a big departure and gives pause for concern. Here are mine so far:

**Less technical, but ultimately more important**

* Devising a programmatic guard for experiments should be independent from settling the necessary debates between producers and consumers of experimental APIs. Said debates must cover expectations, responsibilities and protocols for removing/altering experimental APIs.
* However, I am convinced that the mere introduction of this locking API will put all discussions on the back burner. That's just sociology — [_worse is better_](https://en.wikipedia.org/wiki/Worse_is_better) and all that.

**More technical**

* The basic elements of this PR seem overall good to me: an object reference is the key, a WeakMap is used, anything can be attached to a piece of data for this key, and everything is ultimately mediated by `__dangerousOptInToUnstableAPIsOnlyForCoreModules`.
* There is a layer on top of that: the configuration abstraction. I understand the motivation for it, per [Adam's explanation here](https://github.com/WordPress/gutenberg/pull/46131#discussion_r1055272678), at least on the surface. The use of symbols is clever. However, I wonder if all that is needed, or if we're just getting ahead of ourselves by adding [porcelain](https://git-scm.com/book/en/v2/Git-Internals-Plumbing-and-Porcelain) in the form of `map` and `experiment`.
* A concrete question: since the registration of actions and selectors is mediated by `registerPrivateActionsAndSelectors`, shouldn't that be the component responsible for mapping them to the store? Then there would be no need for the added surface of `map`.
* Do we need to revisit the following explanation you gave me: _"In short, `lock( store, { selectors: {} } )` affects the store but `useSelect( store )` does not return the store object so these utilities make `unlock( useSelect( store ) )` return the private selectors tied to the store."_ Is `configureLockTarget` the right solution to this? Off the top of my head: could there be a simpler aliasing system, where `myWeakMap.get( myAltKey )` simply points to the same data object as `myWeakMap.get( origExperiment )`? As long as it isn't open to race conditions, of course.
* Somewhat related, but less important: I'm on the fence about introducing dedicated domain functions like `createExperiment`. It doesn't mean much to the reader, who will have to jump to the definition to understand that an experiment is just a fresh unique object. Keeping things a little more obvious and closer to the metal might help, even in such details as `const experiment = {}` (could also be named `key` or `reference`).

adamziel commented:
> I am convinced that the mere introduction of this locking API will put all discussions on the back burner.

Can you point to any specific discussions you're worried would be put on the back burner? 

> Said debates must cover expectations, responsibilities and protocols for removing/altering experimental APIs.

I agree that expectations, responsibilities, and protocols may be relevant in relation to some experimental APIs that the Gutenberg plugin wants to publicly expose for feedback and testing are separate from this discussion. To me, it's separate from this PR.

Here we enable sharing internal functions across packages. At the moment, the internal functions are mere implementation details that may be created, removed, and updated without any specific protocols. 

>  Is configureLockTarget the right solution to this? Off the top of my head: could there be a simpler aliasing system, where myWeakMap.get( myAltKey ) simply points to the same data object as myWeakMap.get( origExperiment )? As long as it isn't open to race conditions, of course.

I'm all for simplifying the API and getting rid of `configureLockTarget`, `map` etc. Here's the problem I was trying to solve:

```js
const privateData = {
   actions: { privateAction },
   selectors: { privateSelector }
};
lock( store, privateData );
// ^ Inlined registerPrivateActionsAndSelectors

// Unlocking dispatch() grants access to privateData.actions
unlock( dispatch( store ) ).privateAction();

// Unlocking select() grants access to privateData.selectors
unlock( select( store ) ).privateSelector();
```

Mapping the full private API to the appropriate subset must happen _somewhere_. At the moment, that's `map`. Now I'm thinking we could tightly couple the `experiments` package to the `data` package and make `unlock()` aware of stores. Something along these lines:

```js
function unlock( object ) {
   const privateData = experiments.get( object );
   if ( isDispatchResult( object ) ) {
      return privateData.actions;
   } else if ( isDispatchResult( object ) ) {
      return privateData.selectors;
   }
   return privateData;
}
```

The problem is we'd actually need to return `mapActions( privateData.actions, store )` and `unlock()` has no access to `store`. Well, perhaps `createReduxStore` could privately attach bound `mapActions` and `mapSelectors` functions to the return value of `dispatch()` and `select()`? 🤔  In which case we'd do:

```js
function unlock( object ) {
   const privateData = experiments.get( object );
   if ( isDispatchResult( object ) ) {
      return privateData.mapActions( privateData.actions );
   } else if ( isDispatchResult( object ) ) {
      return privateData.mapSelectors( privateData.selectors );
   }
   return privateData;
}
```

> since the registration of actions and selectors is mediated by registerPrivateActionsAndSelectors, shouldn't that be the component responsible for mapping them to the store? Then there would be no need for the added surface of map.

Ideally, I guess so. Perhaps we don't need that utility at all, though? Since we're reducing the API, registering private selectors and actions could be just this:

```js
lock( store, { selectors: privateSelectors, actions: privateActions } )
```

>  I'm on the fence about introducing dedicated domain functions like createExperiment.

Same – your comment validates the concerns I was having while adding it. If there's a way to get rid of `configureLockTarget` then I'm all for getting rid of this detail as well.

mcsf commented:
> Can you point to any specific discussions you're worried would be put on the back burner?

Concretely, there was that [long Make Core thread](https://make.wordpress.org/core/2022/08/10/proposal-stop-merging-experimental-apis-from-gutenberg-to-wordpress-core/), but IMO it was hard to keep the conversation focused and there were many misconceptions to fight.

More generally, questions to clarify include:

* If a third party chooses to use an experimental API, aren't they responsible for that choice and isn't WordPress unencumbered and free to modify said API? Does this apply equally to APIs gated by `unlock` and APIs that we can't gate yet (e.g. configuration-based APIs, like theme.json or block.json)?
* Can we keep deprecating and removing experimental APIs of the past? How?
* In situations where an experiment isn't trivial to gate, how hard must we try to jump through hoops in order to leverage `unlock`? In other words, how do we weigh the need for gating against the need to develop the right implementation for a particular problem, if that implementation doesn't lend itself well to gating?
* etc.

> > Said debates must cover expectations, responsibilities and protocols for removing/altering experimental APIs.
> 
> I agree that expectations, responsibilities, and protocols may be relevant in relation to some experimental APIs that the Gutenberg plugin wants to publicly expose for feedback and testing are separate from this discussion. To me, it's separate from this PR.

I agree that it should be separate. In an ideal world it would. I'm not suggesting that we block this PR because of it, btw. I'm just pointing out that there is this other non-code aspect of the matter that needs to be worked on, and warning that the mere merging of this PR could further deprioritise it.

adamziel commented:
I explored removing `configureLockTarget` from the public API and got stuck.

The idea I shared earlier convolutes the code and makes the store setup implicit instead of explicit. The selector-binding function still passed from `@wordpress/data` to `@wordpress/experiments` somehow.

Pseudocode without `configureLockTarget`:

```js
// in redux-store.js:
// Lock receives a data in a certain shape that implies a "magic" `unlock()` behavior 
lock( selectors, {
	storeExperiment,
	mapSelectors,
} );

// in experiments.js
// Unlock uses heuristics to correctly map selectors and actions. This could potentially
// be triggered in a wrong context. Also, this behavior is implicit and could get out of sync with
// the redux-store.js
function unlock( object ) {
   const privateData = getPrivateData( object );
   if ( isDispatchResult( object ) && privateData.mapActions ) {
      return privateData.mapActions( unlock( privateData.storeExperiment ).actions );
   } else if ( isSelectResult( object ) && privateData.mapSelectors ) {
      return privateData.mapSelectors( unlock( privateData.storeExperiment ).selectors );
   }
   return privateData;
}
```

Pseudocode with `configureLockTarget`:

```js
// in redux-store.js:
// The configuration function explicitly requires certain keys that are 
// validated and trigger a well-defined behavior.
configureLockTarget( selectors, {
	experiment: storeExperiment,
	map: mapSelectors
} );

// in experiments.js
// mapFunction is explicit and stored in a dedicated storage. There's no chance of
// applying it by mistake.
function unlock( object ) {
   const { privateData, mapFunction } = getPrivateData( object );
   if ( mapFunction ) {
      return mapFunction( privateData);
   }
   return privateData;
}
```

Out of these two, `configureLockTarget` is easier to explain, understand, test, and is less prone to errors.

That being said, I'd still love to get rid of that part of the API and I'm open to any and all ideas.

adamziel committed:
Rebuild docs

adamziel commented:
Here's another idea:

```js
	lock( store, {
		registerPrivateActions: ( privateActions ) =>
			lock( actions, mapActions( /* ... */ ) ),
		registerPrivateSelectors: ( privateSelectors ) => 
			lock( selectors, mapSelectors( /*...*/ ) )
	} );

	unlock( store ).registerPrivateActions({ privateAction })
	unlock( store ).registerPrivateSelectors({ privateSelector })
```

This would reduce the entire public API of `@wordpress/experiments` to just `__dangerousOptInToUnstableAPIsOnlyForCoreModules`.

Here's the problem: in Gutenberg we import _store descriptors_ not _store instances_ so these functions would be unlocked on a _descriptor_ but affect the instances. I have a feeling this can be done and is worth pursuing.

**Edit:** It worked! I used JavaScript proxy objects to bridge the gap between the store instance and the store descriptor.

adamziel committed:
Remove the configureLockTarget utility

adamziel commented on packages/data/src/redux-store/index.js:
```diff
@@ -148,6 +151,14 @@ export default function createReduxStore( key, options ) {
+				actions,
+				new Proxy( privateActions, {
+					get: ( target, prop ) => {
+						return mapActions( privateActions, store )[ prop ];
```
Would be great to only call `mapActions()` once and not on each call. Ditto for `mapSelectors()` below

adamziel replied:
I'll refrain from optimizing this prematurely – `mapActions` is cheap and lazy mapping requires reconciling the already mapped `privateActions` and the current `privateActions`.

noisysocks replied:
Would `memo()` around the `get()` be an easy way to do this?

adamziel replied:
Oh, hm, yes it would! Perhaps something like:

```js
import memoize from 'memize';

// ...

get: memoize( (target, prop) => {
```

It would require crafting a few smart test cases for different before/after scenarios so I won't block this PR on it. If you'd like to play with it and spin a new PR, I'll gladly help with the reviews!

ramonjd commented on packages/experiments/README.md:
````diff
@@ -1,51 +1,85 @@
 ```
 
-The function name communicates that plugins are not supposed to use it. To make double and triple sure, the first argument must be that exact consent string, and the second argument must be a known `@wordpress` package that hasn't opted in yet – otherwise an error is thrown.
+Each package may only opt-in once. The function name communicates that plugins are not supposed to use it. To make double and triple sure, the first argument must be that exact consent string, and the second argument must be a known `@wordpress` package that hasn't opted in yet – otherwise an error is thrown.
````
```suggestion
Each package may only opt in once. The function name communicates that plugins are not supposed to use it.

The function will throw an error if the following conditions are not met:

1. The first argument must exactly match the required consent string: `'I know using unstable features means my plugin or theme will inevitably break on the next WordPress release.'`. 
2. The second argument must be a known `@wordpress` package that hasn't yet opted into `@wordpress/experiments`
```

Optional copy update. Just to make it clear that the first arg must be an exact match of the `requiredConsent` value.

Also optional, because no one cares 😄 "opt in" and "opt out" are the verbs. "opt-in" and "opt-out" are adjectives. 

🍺

adamziel replied:
Oh that's great, thank you!

ramonjd commented on docs/contributors/code/coding-guidelines.md:
```diff
@@ -136,10 +136,317 @@ export { __unstableDoTerribleAwfulAction } from './api';
+// The stable public component is a thin wrapper that calls the
+// experimental component with the experimental features disabled
+export const MyButton = ( { title } ) => 
+    <ExperimentalMyExistingButton title={ title } __experimentalShowIcon={ false } />
```
Should `ExperimentalMyExistingButton` be `ExperimentalMyButton`? Just trying to grok the example...

adamziel replied:
Yes it should, good spot!

ramonjd commented:
Thanks for the work here. I've read through things and the API makes sense to me. I don't have much to add on the implementation sorry, but it sounds like a good idea!

A question: is the plan to adopt this API for experimental features going forward or would we also migrate existing experimental features (that are not destined to go public any time soon)?

adamziel committed:
Update packages/experiments/README.md

Co-authored-by: Ramon <ramonjd@users.noreply.github.com>

dmsnell commented on packages/experiments/README.md:
```diff
@@ -1,51 +1,90 @@
-Expose a new `__experimental` API as follows:
+The function will throw an error if the following conditions are not met:
+
+1. The first argument must exactly match the required consent string: `'I know using unstable features means my plugin or theme will inevitably break on the next WordPress release.'`. 
```
❤️

dmsnell replied:
as much as I love this, I feel like we also need to be careful if we make this step to match that _promise_ with actually removing experimental interfaces and functions.

as someone recently pointed out in a DefinitelyTyped issue, the `__experimental` function arg has been in for a _long_ time, so continuing to call it experimental can feel ingenuine.

this came up with the original discussion about getting rid of experimental APIs. even if we say with our words that they will "quickly" disappear or break, at some point, if we have stabilized that code and built up a number of examples in Core of using them, then we have actually communicated that experimental code isn't, we've communicated something more like, "this stopped being experimental a while ago but we never renamed it"

want to make sure the severity of our words matches the integrity of our actions

adamziel replied:
Agreed! While we seem to be stuck with certain deprecated and experimental APIs, e.g. editor.RichText is used by 255 plugins according to wpdirectory.net search, I hope that the ability to hide functions will catalyze adding and removing experiments at will.

adamziel committed:
Lint

adamziel committed:
Fix typo in coding-guidelines.md

adamziel commented:
> A question: is the plan to adopt this API for experimental features going forward or would we also migrate existing experimental features (that are not destined to go public any time soon)?

@ramonjd It would be absolutely fantastic to "privatize" the unreleased experimental features to keep WordPress 6.2 free of as many new public experimental APIs as possible.

adamziel commented:
I'd like to merge this PR some time soon if noone sees any blockers. I linted and fixed typos in this PR and I expect the checks to turn green. cc @noisysocks @ntsekouras

noisysocks commented:
> > A question: is the plan to adopt this API for experimental features going forward or would we also migrate existing experimental features (that are not destined to go public any time soon)?
> 
> @ramonjd It would be absolutely fantastic to "privatize" the unreleased experimental features to keep WordPress 6.2 free of as many new public experimental APIs as possible.

Yeah thinking we should `lock()` any `__experimental` and `__unstable` APIs that were added in between 6.1 and 6.2 and then from now on: 1) requiring `lock()` instead of `__experimental` or `__unstable`, and; 2) considering any `__experimental` and `__unstable` APIs added before 6.2 as "stable but discouraged". That should be in line with the discussion on Make/Core about experimental APIs.

I'll quickly look at this PR before you merge 😀

noisysocks commented on packages/data/src/test/privateAPIs.js:
```diff
@@ -0,0 +1,193 @@
+			} );
+
+			const privateSelectors = unlock( registry.select( groceryStore ) );
+			expect( privateSelectors.getSecretDiscount ).toEqual(
```
Might be good to assert here that the public selector still exists on the object returned by `unlock()`.

adamziel replied:
They didn't! So good call – I pushed a remedy in 954b353efe.

noisysocks commented on packages/data/src/test/privateAPIs.js:
```diff
@@ -0,0 +1,193 @@
+				setSecretDiscount,
+			} );
+			const privateActions = unlock( registry.dispatch( groceryStore ) );
+			expect( privateActions.setSecretDiscount ).toEqual(
```
Same again here.

noisysocks commented on docs/contributors/code/coding-guidelines.md:
```diff
@@ -136,10 +136,317 @@ export { __unstableDoTerribleAwfulAction } from './api';
+export const { lock, unlock } =
+	__dangerousOptInToUnstableAPIsOnlyForCoreModules(
+		'I know using unstable features means my plugin or theme will inevitably break on the next WordPress release.',
+		'@wordpress/block-editor' // The package opting in
```
It took me a long long time to realise that this is supposed to be the name of the package _calling_ `__dangerousOptInToUnstableAPIsOnlyForCoreModules`, not the name of the package whose APIs you want to access.

I wonder if there's a way to get the package name from an environment variable or something so that it's clearer.

Probably not. Just noting my confusion nonetheless.

adamziel replied:
Good note! I just updated the docs to clarify that.

As for the automation – both arguments could probably be injected by a custom babel plugin. Brownie points – it would obscure the correct usage even further for extenders trying to reverse-engineer it without consulting the docs. Would you be game for exploring it?

noisysocks commented on packages/experiments/src/implementation.js:
```diff
@@ -0,0 +1,187 @@
+) => {
+	if ( ! CORE_MODULES_USING_EXPERIMENTS.includes( moduleName ) ) {
+		throw new Error(
+			`You tried to opt-in to unstable APIs as a module "${ moduleName }". ` +
```
```suggestion
			`You tried to opt-in to unstable APIs as module "${ moduleName }". ` +
```

noisysocks commented on packages/experiments/src/implementation.js:
```diff
@@ -0,0 +1,187 @@
+	}
+	if ( registeredExperiments.includes( moduleName ) ) {
+		throw new Error(
+			`You tried to opt-in to unstable APIs as a module "${ moduleName }" which is already registered. ` +
```
```suggestion
			`You tried to opt-in to unstable APIs as module "${ moduleName }" which is already registered. ` +
```

noisysocks reviewed (approved):
Left a few notes but overall looks pretty good. Think we gotta start using it and see how it goes 👍

ntsekouras reviewed (approved):
Thanks Adam! Let's get this in and lock some new APIs for 6.2 😄

adamziel committed:
Update packages/experiments/src/implementation.js

Co-authored-by: Robert Anderson <robert@noisysocks.com>

adamziel committed:
Update packages/experiments/src/implementation.js

Co-authored-by: Robert Anderson <robert@noisysocks.com>

adamziel committed:
Clarify the meaning of the second argument to __dangerousOptInToUnstableAPIsOnlyForCoreModules

adamziel committed:
Assert that the public selector and actions still exist on the object returned by unlock().

noisysocks commented:
Nice work!

noisysocks commented:
Two notes so far from trying to use these APIs in the context of https://github.com/WordPress/gutenberg/issues/47196:

- It's unfortunate that you don't get type hinting on what's returned from `unlock()`.

  ```js
  import { Component } from '@wordpress/components'; // 'Component' is typed
  const ExperimentalComponent = unlock( Component ); // 'ExperimentalComponent' has type 'any'
  ```

- In Story Book `__dangerousOptInToUnstableAPIsOnlyForCoreModules` is called multiple times assumedly due to hot module loading. This causes an error when navigating to a different component. 

  <img width="1258" alt="Screenshot 2023-01-18 at 14 17 41" src="https://user-images.githubusercontent.com/612155/213075220-02ffc5da-7a4d-4b77-afe3-2968165ca6dd.png">

adamziel commented:
@noisysocks Thank you for reporting! Both problems can be solved – let's discuss in https://github.com/WordPress/gutenberg/pull/47229 and post the solution in this PR later on.

Edit: [Typehinting discussion](https://github.com/WordPress/gutenberg/pull/47229/files#r1073418480)

adamziel commented:
To close the loop:

> It's unfortunate that you don't get type hinting on what's returned from unlock().

The TypeScript issue [can be solved in a few different ways](https://github.com/WordPress/gutenberg/pull/47229#discussion_r1073418480). Since folks are leaning towards locking all the experiments into a single exported `experiments` object, this would work:

```ts
// In the consumer package:
import { unlock } from './lock-unlock';
import { experiments as componentExperiments } from '@wordpress/components';
const { MyComponent } = unlock( componentExperiments );

// In the components package:
type Experiments = {
    MyComponent: typeof ExperimentalMyComponent
};
export const experiments = {} as HasLockedType< Experiments >;
lock(experiments, {
    MyComponent: ExperimentalMyComponent
});

// In the Experiments package:
type HasLockedType<Self, LockedType> = Self & { LockedType: LockedType };
type UnlockReturn< Arg > =
	Arg extends {LockedType: infer T}
		? T
		: Arg;
function unlock<T>( arg: T ) : UnlockReturn<T> { /* ... */ }
```

> In Story Book __dangerousOptInToUnstableAPIsOnlyForCoreModules is called multiple times assumedly due to hot module loading. This causes an error when navigating to a different component.

@noisysocks's PR adds a new `ALLOW_EXPERIMENT_REREGISTRATION` feature flag that allows multiple registrations of the same package outside of WordPress core.

mcsf commented:
@adamziel, I just wanted to say that I quite like how this turned out, with the reduced surface and the proxy-powered unlocking.
